### DISCUSSION TITLE
Noob cannot scan drum snippet

### DISCUSSION THREAD

**Discussion by infojunkie** [Audiveris/audiveris#672] [Q&A]
I'm trying to convert the attached [drum notation from Wikipedia](https://en.wikipedia.org/wiki/Fandango). I get the message:

![Fandango_dance_pattern](https://github.com/Audiveris/audiveris/assets/92977/5d71c3fa-239c-4659-bf2a-7b79ee40dc4e)

Fandango_dance_pattern With an interline value of 9 pixels, either this sheet contains no staves, or the picture resolution is too low (try 300 DPI).

Is there an option setting to help with this? Thanks!

**Comment by hbitteur** [comment]
@infojunkie 
One of the very first tasks of Audiveris engine is the `SCALE` step.
This step tries to determine the global scaling of the input sheet, especially the "_interline value_" which is the mean vertical distance between staff lines. This interline value, generally around 20 pixels, is key for the following processing steps.

This value is measured **before** the staff lines detection, simply by looking at the histogram of vertical runs.
See details in [this section](https://audiveris.github.io/audiveris/_pages/main/sheet_scale/) of Audiveris handbook.

In your example, there is just no vertical distance to measure! Therefore, the OMR engine got lost.

I just tried to manually force the scaling data (using the menu `Sheet | Set Scaling Data`), but then the engine crashed in the following step (`GRID`) which retrieves the staves and the vertical barlines, here again driven by the (here unknown) staff height.

I would need to further investigate to make the engine survive this uncommon case.
But is it worth the effort? (knowing that I have several on-going tasks right now...)

**Comment by infojunkie** [comment]
Thanks for your reply. I will try to replicate your steps.

Why is this case uncommon ? Is it because there's only a single line? It is a common case if you consider music snippets in text books.

**Comment by hbitteur** [comment]
I said "uncommon" because Audiveris is focused on the processing of sheets of music.
Here we have only a snippet, which means less information the engine can rely upon.

OK, I will try to make it work (in some way) on such snippets.

**Comment by infojunkie** [comment]
Thanks for your quick replies - much appreciated!

**Comment by hbitteur** [comment]
I have been able to process your snippet, after some manual entries and a few modifications in the code to fit your specfic case.
I can't release the code at it is, because it will no longer work on the other 1-line staff examples I have seen so far.

Here is an example:

![image](https://github.com/Audiveris/audiveris/assets/22053634/2ebd7b09-838f-40bb-8e1c-bbfb71df1f79)

As you can see, we have a system composed of an upper staff of 5 lines and a lower staff of 1 line.
Both staves exhibit the same barline height (4 interlines)

In your case, the barline height is about half the size of these (2 interlines)

I also saw cases where the lower staff started with a half barline (as if the part surrounded by a green circle in the picture above did not exist). Unfortunately I can't immediately put my hands on a concrete example.

We need a way to tell the OMR engine what the expected barline height is for the 1-line staves in the score at hand.
Either:
- 4 interlines for all barlines in staff
- 2 interlines for all barlines in staff
- 2 interlines above the staff for the initial barline, then 4 interlines for the following barlines

How could the user enter this kind of information? Please help me.

**Comment by infojunkie** [comment]
If you need the user to describe the features of the score, then I imagine that the more detailed the description, the easier the engine will be able to perform the scan. Maybe it would be worth designing a DSL that describes the layout of the score. For example, can the description be expressed in MusicXML? The user would supply the "template" of the score(s) they are sending to the engine.

**Comment by hbitteur** [comment]
I'm afraid asking for some DSL would be too complex for everyone, beginning by the end user.

I think the best entry point is the dialog `Book | Set Book Parameters`.
We need to augment it with information about the formatting of these 1-line staves (at sheet, book or global levels).
Once we have defined precisely which information is needed, the rest is just a matter of UI design.

Before the question of barline heights for 1-line staves, we must tell the engine not to try to measure any interline value.

![image](https://github.com/Audiveris/audiveris/assets/22053634/8413bfcb-c977-4584-8a24-d4d6cd8bbb43)

Perhaps, we could do this simply via an additional processing switch, like "5-line standard staves", which would be true by default.
By setting this switch to false, together with the other multi-line staves switches which are false by default, the user would explicitly tell the engine that there is no multi-line staff in the current sheet (or book). And thus prevent the SCALE step to look for a concrete interline.

If the user checks the "1-line percussion staves" to true, the dialog could display additional  field(s) about barline heights, specified in number of interlines.
But still the engine will need an interline value (specified in pixels). This is something the user can manually enter via the menu `Sheet | Set Scaling Data`, but this must be done right after the SCALE step, before the GRID step. Not very convenient.
